keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+actions: https://github.com/kubernetes/kubeadm/issues/1949
+2) Copy kubeadm to the k/kubeadm repository.
+3) Implement build / hack / verify tooling for k/kubeadm.
+4) Add pre-submit e2e jobs for k/kubeadm. Ideally there should be
+two per branch - one that verifies / builds the code base
+and one that does a full e2e using "kinder".
+5) Implement tooling in the k/release repository to build kubeadm artifacts.
+6) Implement a post-submit job that will sync the k/kubeadm repository
+tags and branches with those of k/k. Keep the job dry-running as a start.
+7) Adapt the reference documentation tooling maintained by SIG Docs to build
+from the kubeadm repository.
+8) Announce the move on the kubernetes-dev mailing list.
+9) Remove kubeadm and references to cmd/kubeadm from k/k.
+10) Enable the post-submit job (disable its dry-run), so that new k/k
+tags and branches can be synced in k/kubeadm.
+
+### User Stories
+
+#### Story 1
+
+As a user, I would like to build kubeadm without depending on the
+kubernetes/kubernetes repository.
+
+kubeadm no longer imports k/k packages, so this is already possible
+if one just copies cmd/kubeadm outside of k/k and creates a new Go
+module for it. It still requires some hacks around applying a version in the
+kubeadm binary. Once kubeadm moves out both the build tooling and module
+will be available to users for easy building.
+
+#### Story 2
+
+As a user, I would like to import the kubeadm public, versioned API without
+importing kubernetes/kubernetes.
+
+Currently importing the k/k root module will make your build fail.
+The best solution for users is to fork the kubeadm API packages.
+Once kubeadm moves out, importing the "kubeadm" module will make
+it possible to easily import only the API portion.
+
+#### Story 3
+
+As a user, I would like to import utility packages from kubeadm.
+
+Similar to the above user story, importing kubeadm from k/k will fail.
+Once kubeadm moves out, utility functions will be exposed for public use,
+but perhaps without any guarantees until the kubeadm backend is truly
+announced for public consumption.
+
+### Implementation Details/Notes/Constraints
+
+#### Location
+
+The proposed new location is the kubernetes/kubeadm repository.
+
+This makes it possible to use `go get k8s.io/kubeadm` to obtain
+the kubeadm source tree.
+
+The current contents of the `k/k/cmd/kubeadm/app` directory
+should be moved in the root of k/kubeadm repository.
+
+The current `k/k/cmd/kubeadm/test` directory that contains
+integration tests should be merged with the `k/kubeadm/tests`
+directory under `k/kubeadm/tests/integration`.
+
+Existing tools in the k/kubeadm repository such as "kinder" and
+"operator" must continue existing as separate modules (sub-folders)
+inside the kubeadm module and can be obtained using
+`go get k8s.io/kubeadm/some-tool`.
+
+Existing docs and tests must co-exist in sub-folders too.
+
+Side effects:
+- All contents of k/kubeadm will be branched and tagged with kubeadm,
```

Comment:
"tagged with kubeadm" is confusing.  It reads as if "kubeadm" is the tag.  I suspect what is meant is more "All contents of k/kubeadm will be branched and tagged in a semver compliant fashion..."

Reply:
i'm going to update this.